### PR TITLE
Custom DSL parser prototype 

### PR DESCRIPTION
**Pull request by chrisspre**
Initial parser for custom DSL

Very rough draft of a parser for an unfinished DSL design, written in C# for now.

Please provide feedback on

the rough direction of the DSL
todays discussion about actions/functions and annotations is not included.
folder layout
at the moment everything is under \src\custom-sdl\dotnet
It translates:

enum employmentType { salaried hourly }

type company
{
    @key stockSymbol: string
    name: string
    incorporated: dateTime
    employees: [employee]

    test: [integer]
    ceo: employee
}

type employee
{
    @key id: integer 
    key: boolean
    name : name    
    title: string
    employmentType: employmentType

    // a CSDL function
    promote() : integer     
    
    // a CSDL function overload
    promote(levels: integer, note:string) : integer 

    // a CSDL action with a required parameter that can be null
    @action fire(date: dateTime?) 
}

type name
{
    firstName : string
    middleInitial: string?
    lastName: string
    title: string?
}
to

    <Schema Namespace="rapid" xmlns="http://docs.oasis-open.org/odata/ns/edm">
      <EnumType Name="employmentType">
        <Member Name="salaried" Value="0" />
        <Member Name="hourly" Value="1" />
      </EnumType>
      <EntityType Name="company">
        <Key>
          <PropertyRef Name="stockSymbol" />
        </Key>
        <Property Name="stockSymbol" Type="Edm.String" Nullable="false" />
        <Property Name="name" Type="Edm.String" Nullable="false" />
        <Property Name="incorporated" Type="Edm.DateTimeOffset" Nullable="false" />
        <Property Name="test" Type="Collection(Edm.Int32)" Nullable="false" />
        <NavigationProperty Name="employees" Type="Collection(rapid.employee)" />
        <NavigationProperty Name="ceo" Type="rapid.employee" Nullable="false" />
      </EntityType>
      <EntityType Name="employee">
        <Key>
          <PropertyRef Name="id" />
        </Key>
        <Property Name="id" Type="Edm.Int32" Nullable="false" />
        <Property Name="key" Type="Edm.Boolean" Nullable="false" />
        <Property Name="name" Type="rapid.name" Nullable="false" />
        <Property Name="title" Type="Edm.String" Nullable="false" />
        <Property Name="employmentType" Type="rapid.employmentType" Nullable="false" />
      </EntityType>
      <ComplexType Name="name">
        <Property Name="firstName" Type="Edm.String" Nullable="false" />
        <Property Name="middleInitial" Type="Edm.String" />
        <Property Name="lastName" Type="Edm.String" Nullable="false" />
        <Property Name="title" Type="Edm.String" />
      </ComplexType>
      <Function Name="promote" IsBound="true" IsComposable="true">
        <Parameter Name="this" Type="rapid.employee" Nullable="false" />
        <ReturnType Type="Edm.Int32" Nullable="false" />
      </Function>
      <Function Name="promote" IsBound="true" IsComposable="true">
        <Parameter Name="this" Type="rapid.employee" Nullable="false" />
        <Parameter Name="levels" Type="Edm.Int32" Nullable="false" />
        <Parameter Name="note" Type="Edm.String" Nullable="false" />
        <ReturnType Type="Edm.Int32" Nullable="false" />
      </Function>
      <Action Name="fire" IsBound="true">
        <Parameter Name="this" Type="rapid.employee" Nullable="false" />
        <Parameter Name="date" Type="Edm.DateTimeOffset" />
      </Action>
    </Schema>

### CODE REVIEW

File: tools/rsdl/dotnet/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# demo
+
+the demo program can be run like
+
+```sh
+dotnet run -- --input jetsons.rsdl --format JSON
```
Comment (wtrocki) [inline tools/rsdl/dotnet/README.md:6]: Would it be possible to:

- Add information how to compile program etc.
- rename output file to rapid etc.

File: tools/rsdl/dotnet/demo/Program.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
+﻿using Newtonsoft.Json;
```
Comment (wtrocki) [inline tools/rsdl/dotnet/demo/Program.cs:1]: Package name needs to be changed - it is `demo`

**Comment by wtrocki** [comment]
I have tried it and there are couple things around error handling and validation that can be improved in follow up PR's. For the moment we might bring that to the repository. Then I will help to clean this up and remove old ideas (also add build process and some testing for this in an early stage!

Amazing work!

File: tools/rsdl/dotnet/README.md (hunk 1 above)
Comment (ralfhandl) [reply tools/rsdl/dotnet/README.md:11]: I got it running by following all hints in VS Code, e.g. the ones mentioned above.

Would be nice if someone could verify from scratch and complete the description.

**Review by ralfhandl — APPROVED** [review]
Could build and run it following the hints in VS Code plus some guessing.

Added the correct guesses to `dotnet/README.md`

File: tools/rsdl/syntax-highlighting/vsc-extension-quickstart.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+* This folder contains all of the files necessary for your extension.
+* `package.json` - this is the manifest file in which you declare your language support and define the location of the grammar file that has been copied into your extension.
+* `syntaxes/rsdl.tmLanguage.json` - this is the Text mate grammar file that is used for tokenization.
+* `language-configuration.json` - this is the language configuration, defining the tokens that are used for comments and brackets.
+
+## Get up and running straight away
+
+* Make sure the language configuration settings in `language-configuration.json` are accurate.
+* Press `F5` to open a new window with your extension loaded.
```
Comment (ralfhandl) [inline tools/rsdl/syntax-highlighting/vsc-extension-quickstart.md:13]: @chrisspre Apparently the necessary launch configuration is missing: 
- [ ] add to project
- [ ] mention here

**Review by ralfhandl — APPROVED** [review]

**Comment by wtrocki** [comment]
Amazing job @ralfhandl
Got recommendation for extension when I got the source code again and compiled this.

**Review by wtrocki — APPROVED** [review]

File: tools/rsdl/syntax-highlighting/vsc-extension-quickstart.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Welcome to your VS Code Extension
```
Comment (wtrocki) [inline tools/rsdl/syntax-highlighting/vsc-extension-quickstart.md:1]: Actually need to test that as well. at some point extension can be added to marketplace and documented on website.

File: tools/rsdl/dotnet/demo/Program.cs (hunk 2 above)
Comment (chrisspre) [reply tools/rsdl/dotnet/demo/Program.cs]: change the folder name to demo, the name of the binary is not controlled by that (in .net)

**Pull request merged**